### ISSUE TITLE
Support fetching results in batches in MSSQL engine

### ISSUE THREAD

**Issue by inglesp** [opensafely-core/databuilder#531]
To work around the flakiness of the database connection, the MSSQL engine should write the results to a temporary table, and then download the results in batches.

This involves porting [this code from the legacy MSSQL engine](https://github.com/opensafely-core/databuilder/blob/4a49fc7c72fdf1d06ae75ad01ec91455ad22749c/databuilder/query_engines/legacy_mssql.py#L49-L81) to the [new MSSQL engine](https://github.com/opensafely-core/databuilder/blob/main/databuilder/query_engines/mssql.py).

**Comment by evansd** [comment]
> Note that the legacy query engines had a .get_queries() method, while the new engine has a .get_query() method, so the implementation of [fetch_results_in_batches()](https://github.com/opensafely-core/databuilder/blob/4a49fc7c72fdf1d06ae75ad01ec91455ad22749c/databuilder/query_engines/mssql_lib.py#L29-L111) will have to change.

Ah, I'm actually planning to restore the original API because we'll need it for #515. I was planning to defer that change until it was actually needed, but it would be very easy to change the API now and add some comments explaining why it needs to be that way.

**Comment by evansd** [comment]
In cohortextractor we went to some effort to deterministically generate the temporary table name from the query, and then to check whether an appropriately named table already exists before trying to write results to it.

This gave us recoverability in the sense that if a download failed the user could re-run the job and, instead of having to re-run all the queries, it would pick up the existing results table and start downloading that. This was very helpful in it's time.

However, it has some downsides:

1. It adds extra complexity to the code. (Even things like checking whether the table exists aren't straightforward because we don't have full introspection permissions on SQL Server.)

2. Attempting to run the same query multiple times in parallel triggers an error. I used to think this wouldn't be a case worth worrying about. But it turns out it happens more often than I expected, particularly in the early days of a study's development. Usually it's the result of someone doing something unintended, but the error message is not at all helpful to the user.

And now that we have better [backoff and retry](https://github.com/opensafely-core/cohort-extractor/pull/774) logic I would expect the cases where the old behaviour would be helpful to be quite limited.

So I'm proposing **not** to reimplement this logic in databuilder. The results tables can just be named something like `results_YYYYMMDD_<unique_random_string>`.

If we do get a case where someone's spent a week generating results that are in a temporary table but failed to download then we could do something by hand to retrieve them.
